server/schedule/operator/builder.go

```diff
@@ -285,74 +428,106 @@ func (b *Builder) buildSteps(kind OpKind) (OpKind, error) {
+	b.currentPeers.Set(peer)
+	delete(b.toPromote, peer.GetStoreId())
 }
 
-func (b *Builder) execAddPeer(p *metapb.Peer) {
+func (b *Builder) execDemoteFollower(peer *metapb.Peer) {
+	b.steps = append(b.steps, DemoteFollower{ToStore: peer.GetStoreId(), PeerID: peer.GetId()})
+	b.currentPeers.Set(peer)
+	delete(b.toDemote, peer.GetStoreId())
+}
+
+func (b *Builder) execAddPeer(peer *metapb.Peer) {
 	if b.isLightWeight {
-		b.steps = append(b.steps, AddLightLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+		b.steps = append(b.steps, AddLightLearner{ToStore: peer.GetStoreId(), PeerID: peer.GetId()})
 	} else {
-		b.steps = append(b.steps, AddLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
-	}
-	if !core.IsLearner(p) {
-		b.steps = append(b.steps, PromoteLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+		b.steps = append(b.steps, AddLearner{ToStore: peer.GetStoreId(), PeerID: peer.GetId()})
 	}
-	b.currentPeers.Set(p)
-	if b.peerAddStep == nil {
-		b.peerAddStep = make(map[uint64]int)
+	if !core.IsLearner(peer) {
+		b.steps = append(b.steps, PromoteLearner{ToStore: peer.GetStoreId(), PeerID: peer.GetId()})
 	}
-	b.peerAddStep[p.GetStoreId()] = len(b.steps)
-	b.toAdd.Delete(p.GetStoreId())
+	b.currentPeers.Set(peer)
+	b.peerAddStep[peer.GetStoreId()] = len(b.steps)
+	delete(b.toAdd, peer.GetStoreId())
 }
 
-func (b *Builder) execRemovePeer(p *metapb.Peer) {
-	b.steps = append(b.steps, RemovePeer{FromStore: p.GetStoreId()})
-	b.currentPeers.Delete(p.GetStoreId())
-	b.toRemove.Delete(p.GetStoreId())
+func (b *Builder) execRemovePeer(peer *metapb.Peer) {
+	b.steps = append(b.steps, RemovePeer{FromStore: peer.GetStoreId(), PeerID: peer.GetId()})
+	delete(b.currentPeers, peer.GetStoreId())
+	delete(b.toRemove, peer.GetStoreId())
 }
 
-// check if a peer can become leader.
-func (b *Builder) allowLeader(peer *metapb.Peer) bool {
-	if peer.GetStoreId() == b.currentLeader {
+var stateFilter = filter.StoreStateFilter{ActionScope: "operator-builder", TransferLeader: true}
+
+// check if the peer has the ability to become a leader.
+func (b *Builder) hasAbilityLeader(peer *metapb.Peer) bool {
+	// these roles are not allowed to become leaders.
+	switch peer.GetRole() {
+	case metapb.PeerRole_Learner, metapb.PeerRole_DemotingVoter:
+		return false
+	}
+
+	// store does not exist
+	if peer.GetStoreId() == b.currentLeaderStoreID {
 		return true
 	}
-	if core.IsLearner(peer) {
+	store := b.cluster.GetStore(peer.GetStoreId())
+	return store != nil
+}
+
+// check if the peer is allowed to become the leader.
+func (b *Builder) allowLeader(peer *metapb.Peer) bool {
+	// these roles are not allowed to become leaders.
+	switch peer.GetRole() {
+	case metapb.PeerRole_Learner, metapb.PeerRole_DemotingVoter:
 		return false
 	}
+
+	// store does not exist
+	if peer.GetStoreId() == b.currentLeaderStoreID {
+		return true
+	}
 	store := b.cluster.GetStore(peer.GetStoreId())
 	if store == nil {
 		return false
 	}
```

Comment:
It seems the logic here is as same as the `hasAbilityLeader`, we can directly call this function here.

Reply:
If `hasAbilityLeader` return `true`, I don't know if it is because `peer.GetStoreId() == b.currentLeaderStoreID` or `store != nil`.